Tx-state aware grouped traversal

fickludd opened the pull request:
Complete transaction state awareness in Kernel API traversal. This PR add support for tx-state awareness when traversing via `NodeCursor.relationships( RelationshipGroupCursor )` and `NodeCursor.relationshipGroupReference`. Note that the `RelationshipGroupCursor.*count()` methods on are still not tx-state aware. We have a card for this.

Builds on #10841. First commit is `1184f3a`. The ivory tower goes higher.

fickludd committed:
Refactor references to allow more compact encoding

References now use the highest bit as the FLAG_MARKER, and
the following 3 bits to specify the encoding. This means that
any reference type can support 1 un-encoded mode and 8 encoded
modes, with 60 bits remaining for id-space.

fickludd committed:
Add encodeTxStateFiltering relationship reference

fickludd committed:
Low-level added relationship methods tx-state node

fickludd committed:
Remove RelationshipDirection.ALL, add RelationshipDirection.ERROR

fickludd committed:
Tx-state aware traversal by group

fickludd committed:
Refactor references: turn relationship reference encoding into enum

fickludd committed:
checkstyle

tinwelint commented on community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/NodeState.java:
```diff
@@ -59,5 +60,9 @@ void visitLabelChanges( long nodeId, Set<Integer> added, Set<Integer> removed )
 
     PrimitiveLongIterator getAddedRelationships( Direction direction, int[] relTypes );
 
+    PrimitiveLongIterator getAddedRawRelationships();
```
What is "raw relationships"? Please take the opportunity to describe that in javadocs.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/RelationshipDirection.java:
```diff
@@ -0,0 +1,47 @@
+ *  a.getRelationships( org.neo4j.graphdb.Direction.INCOMING ) => r2, r3
+ *  a.getRelationships( org.neo4j.graphdb.Direction.BOTH ) => r1, r2, r3
+ */
+public enum RelationshipDirection
```
Yey, our third direction enum! I'm guessing the `storageengine` direction will go away soon though, or hopefully at least.

fickludd replied:
I don't really see the point of the `storageengine` one, as it's a copy of the `graphdb` direction. This one at least has different (and described) semantics. But yeah, more enums can't be bad, right?

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/RelationshipDirection.java:
```diff
@@ -31,7 +31,6 @@
  *  a.getRelationships( OUTGOING ) => r1
  *  a.getRelationships( INCOMING ) => r2
  *  a.getRelationships( LOOP ) => r3
- *  a.getRelationships( ALL ) => r1, r2, r3
```
Would be great w/ explanation of what `ERROR` is here

tinwelint commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java:
```diff
@@ -299,4 +280,113 @@ private void traverseWithoutGroups( RelationshipTestSupport.StartNode start, boo
+            expectedCounts.put( kv.getKey(), rs.size() + 1 );
+        }
+
+        String newLabelName = "NEW";
```
I know I'll probably push your buttons here, but can we try to limit `label` for relationship type to a minimum? It looks especially weird since the methods has the term `relationshipType`. A variable called something... label primes the mind that it is a node label and you'll have to look at usages to figure out whether it's a node label or relationship type.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/RelationshipTraversalCursor.java:
```diff
@@ -82,6 +82,30 @@ boolean check( long source, long target, long origin )
+            case ERROR:
+                throw new IllegalArgumentException( "There has been a RelationshipDirection.ERROR" );
+            default:
+                throw new IllegalStateException( "Still poking my eye, dear checkstyle..." );
```
Cute, but something more informative, like including `direction` too?

tinwelint commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java:
```diff
@@ -193,30 +200,6 @@ public void shouldSeeRelationshipInTransactionBeforeCursorInitialization() throw
 
-//    @Test
-//    public void shouldTraverseSparseNodeViaGroups() throws Exception
-//    {
```
Some left-overs

tinwelint commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTestSupport.java:
```diff
@@ -25,7 +25,11 @@
-import org.neo4j.graphdb.*;
+import org.neo4j.graphdb.Direction;
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Node;
```
Time to left IntelliJ know about this rule?

fickludd replied:
Strange thing is I never got that before, and fixed it with `optimize imports`, so I think IntelliJ knows. Dunno if I somehow managed to switch to default code style for a period when that happened...

fickludd committed:
address review comments

 - change relationship label to relationship type
 - remove raw from TransactionState NodeState.getAddedRawRelationships
 - improve checkstyle blaming exception message

tinwelint reviewed (approved):
You still have some commented tests in `RelationshipTransactionStateTestBase.java`, but I'll approve anyway
